src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +112,226 @@ public class FromBase64Transform : ICryptoTransform
+
+            if (_inputBuffer == null)
+                ThrowHelper.ThrowObjectDisposed();
 
-            int effectiveCount;
-            byte[] temp = GetTempBuffer(inputBuffer, inputOffset, inputCount, out effectiveCount);
+            if (outputBuffer == null)
+                ThrowHelper.ThrowArgumentNull(ThrowHelper.ExceptionArgument.outputBuffer);
 
-            if (effectiveCount + _inputIndex < 4)
+            // The common case is inputCount = InputBlockSize
+            byte[] tmpBufferArray = null;
+            Span<byte> tmpBuffer = stackalloc byte[InputBlockSize];
+            if (inputCount > InputBlockSize)
             {
-                Buffer.BlockCopy(temp, 0, _inputBuffer, _inputIndex, effectiveCount);
-                _inputIndex += effectiveCount;
+                tmpBuffer = tmpBufferArray = CryptoPool.Rent(inputCount);
+            }
+
+            tmpBuffer = GetTempBuffer(inputBuffer.AsSpan(inputOffset, inputCount), tmpBuffer);
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
+
+            // Too little data to decode: save data to _inputBuffer, so it can be transformed later
+            if (bytesToTransform < Base64InputBlockSize)
+            {
+                tmpBuffer.CopyTo(_inputBuffer.AsSpan(_inputIndex));
+
+                _inputIndex = bytesToTransform;
+
+                ReturnToCryptoPool(tmpBufferArray, tmpBuffer.Length);
+
                 return 0;
             }
 
-            byte[] result = ConvertFromBase64(temp, effectiveCount);
+            ConvertFromBase64(tmpBuffer, outputBuffer.AsSpan(outputOffset), out int consumed, out int written);
+            Debug.Assert(consumed == bytesToTransform);
 
-            Buffer.BlockCopy(result, 0, outputBuffer, outputOffset, result.Length);
+            ReturnToCryptoPool(tmpBufferArray, tmpBuffer.Length);
 
-            return result.Length;
+            return written;
         }
 
         public byte[] TransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount != InputBlockSize is allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+
+            if (_inputBuffer == null)
+            {
+                ThrowHelper.ThrowObjectDisposed();
+            }
+
+            if (inputCount == 0)
+            {
+                return Array.Empty<byte>();
+            }
 
-            int effectiveCount;
-            byte[] temp = GetTempBuffer(inputBuffer, inputOffset, inputCount, out effectiveCount);
+            // The common case is inputCount <= Base64InputBlockSize
+            byte[] tmpBufferArray = null;
+            Span<byte> tmpBuffer = stackalloc byte[Base64InputBlockSize];
+            if (inputCount > Base64InputBlockSize)
```

Comment:
Similar "would stackalloc more here hurt? would it help?" as above.

Reply:
Same as https://github.com/dotnet/corefx/pull/38062#discussion_r301557728